lib/netext/httpext/response.go

```diff
@@ -136,6 +137,8 @@ func (res *Response) JSON(selector ...string) (interface{}, error) {
 			body = b
 		case string:
 			body = []byte(b)
+		case goja.ArrayBuffer:
+			body = b.Bytes()
```

Comment:
I am definitely against `httpext` depending on `goja` ;). I also don't think that supporting neither `[]byte` or `goja.ArrayBuffer` makes a lot of sense here, so I am for just dropping both.

Reply:
True, I didn't want to refactor this to remove the dependency, and also think it doesn't make sense to support this use case.

That said, dropping `[]byte` would be yet another breaking change (we have a test for this use case: `TestNoResponseBodyMangling`), and practically I'm not sure convoluting the code just to remove the dependency is worth it.

@na-- Do you agree with dropping both `[]byte` and `goja.ArrayBuffer` from here?

Reply:
Hmm I completely agree with @MStoykov that we shouldn't depend on goja in httpext, but I am not sure we can drop `[]byte` from here - what would happen if the user has `responseType: binary` and tries to use `.json()`? :confused: 

Rather, maybe we should just move the whole .JSON method in there? https://github.com/loadimpact/k6/blob/a379c3787372ce2a4c70cb2049fc43be7af798a2/js/modules/k6/http/response.go#L47

Reply:
>  `responseType: binary` and tries to use `.json()` 

Why would they do that? 

I would argue that this just doesn't seem like a case that we need to support or something that anyone would've done. So between, making this PR more complicated and moving stuff, and just .. dropping a switch, I much prefer the second one, given that I don't think it has ever been exercised

Reply:
> Why would they do that?

Why not? It will actually be more performant since goja won't do the utf8->utf16 conversion... And why would wanting the response type to be binary preclude you from accessing it as a JSON? 

I am fine with making minor breaking changes that are useful for the vast majority of users, but as far as I can see, the reason for this proposed breaking change is "it will be slightly easier to keep the k6 code a bit cleaner"... Not good enough of a reason in my book, sorry...

If you dislike the move of the `.json()` method back to the JS struct, here's an alternative that has some other performance benefits. Always use `string` or `[]byte` in `httpext.Response.Body`, and use a `Proxy` in the JS `Response` wrapper for accessing it. That proxy will return a `goja.ArrayBuffer` if the response type is `binary`, but it will also lazily return string if the response type is text. This will prevent a bunch of unnecessary utf8->utf16 conversions, right?

Reply:
> Why not? It will actually be more performant since goja won't do the utf8->utf16 conversion... And why would wanting the response type to be binary preclude you from accessing it as a JSON?

I don't think a single user would know that and would have done it. In the same way that I don't think that if a user will use `.json()` they would've before that set `responseType: binary`. So for me, this seems like a "breaking change" that will not be hit by anyone. 

Given that and the fact that the benefit from this being supported is based on users doing stuff and will require some amount of additional code that isn't really needed, I am for dropping this now and either having a PR to reimplement it later, maybe as part of #946. 

Or better (IMO) have the `response.body` field be a [property accessor](https://pkg.go.dev/github.com/dop251/goja#Object.DefineAccessorProperty) that only transforms to `string` or `ArrayBuffer` when required and if that isn't called we can optimize to not actually do any transformation and just work on `[]byte` (or however we save it). 

But my argument is that this is very unlikely to have been used and this PR has enough things that can break without any amount of additional complicated or not complicated refactoring for features (I  argue) that aren't used.

Reply:
I agree with Mihail that it's unlikely someone would deliberately use `responseType: 'binary'` and then called `res.json()`, but they might've done it inadvertently so dropping support for it now would likely break some scripts.

That said, since Mihail suggested we drop `[]byte` from everywhere where `ArrayBuffer` is used now, my vote is to drop it from here as well, and test this thoroughly with existing scripts (canary) and consider bringing it back with either the Proxy or property accessor approach if needed.

Reply:
I still don't see any benefits for users mentioned as a reason for the breaking change, and I see easy workaround that would prevent the breaking change, with performance side-benefits on top, so my position is still that we shouldn't make a breaking change... :slightly_smiling_face: 

> Or better (IMO) have the `response.body` field be a [property accessor](https://pkg.go.dev/github.com/dop251/goja#Object.DefineAccessorProperty) that only transforms to `string` or `ArrayBuffer` when required and if that isn't called we can optimize to not actually do any transformation and just work on `[]byte` (or however we save it).

This seems awesome and _much_ easier to do compared to a `Proxy`, so what's the drawback of just doing that?

Reply:
The breaking change wouldn't have benefits to users, of course. The reason is more to remove a feature that didn't make sense before and to avoid any complicated refactors.

But I also like the property accessor approach and am currently exploring it, so will push a change soon-ish.

Reply:
So I added the dynamic `body` property via the `DefineAccessorProperty` method but it presents some challenges which I don't think are worth overcoming.

The way to set it up involves embedding `goja.Object` in the `httpext.Response` wrapper:

```go
type Response struct {
	*httpext.Response `js:"-"`
	*goja.Object      `js:"-"`
}
```

The problem with this is that it essentially overrides all `httpext.Response` fields, so properties like `status`, `cookies`, etc. return `undefined`. To fix it we would have to use `Proxy` or `DynamicObject` instead, which would involve some reflection for things like `Keys()` or `OwnKeys()`.

We would also need to add caching for the dynamic `body` property since there's no point in re-evaluating it given that the `responseType` remains static.

Maybe I'm not implementing this properly and we wouldn't need to jump through all these hoops, so please let me know if there's a simpler way. But if this is the only way then I'd have to push back as the added complexity of goja-fying `Response` doesn't seem worth it.

Reply:
an "easier" way (that I haven't tested) will be to 
```
val := rt.ToValue(responseObject)
obj := val.ToObject(rt) // which should return a goja object
obj.DefineAccessorProperty ... // do it here, I expect you will not be able to overwrite names 
```

If that doesn't work I guess you can make an Object and set its prototype to the Response object ... which though is probably wrong in some other aspect. 

FYI source for the [objectGoReflect](https://github.com/dop251/goja/blob/ed7b3dc62eb92d6cf9e1782b3a2595e21c851819/object_goreflect.go#L79)

Reply:
Thanks, but the first approach of defining the accessor on `Response` just doesn't call the getter so it doesn't work.

The second one with the prototype does work, and fields like `status` and `cookies` are accessible, but it has other issues. See https://github.com/imiric/k6/commit/017270f8c501ec940abceff3267392c28ad1528a

For one, we'd need to deal with `*goja.Object` internally instead of `*Response`, which is a PITA.

But more importantly, it doesn't passthrough the methods on `Response`, so `.json()` or `.html()` don't work. Calling `Object.keys()` also returns just `["body"]`.

So it seems like `Proxy` or `DynamicObject` would be the way to go, but I'm not sure we should.

Also `DefineDataProperty` seems like a better fit in this case since it avoids any caching requirements.

Reply:
at 

https://github.com/imiric/k6/blob/017270f8c501ec940abceff3267392c28ad1528a/js/modules/k6/http/response.go#L56-L58

You need to provide a pointer to the http.Response as the JSON method is defined on the pointer receiver ;)

Reply:
Thanks, that takes care of it then, but I'm still not a fan of `*goja.Object` being passed around everywhere. :-/

Anyway, I'll clean this up and push it here.

Reply:
I did try to prevent us from having to refactor all of this for a "use-case" that doesn't exist, but :man_shrugging:

Reply:
OK, see the changes for this in 412cbd10, but frankly, it's awful...

- Turns out we can't use `DefineDataProperty` since response bodies for `http.batch()` requests are lazily evaluated, so I had to switch back to `DefineAccessorProperty`.

- I considered adding a cache for `body`, but that would mean doubling the memory usage (it would be stored both as `[]byte` on `httpext.Response.Body` and as `string` or `ArrayBuffer` in the cache). I guess without it we technically still get the memory penalty, and also the CPU overhead of needlessly calling `getBody()` everytime it's accessed. Or... we could just go back to storing it once in `httpext.Response.Body` as `interface{}`.

- `Object.keys(resp)` still returns only `["body"]`. For this to work I think we'd need to switch to `Proxy` or `DynamicObject`, which has its own challenges as I commented above.

- `*goja.Object` is passed around internally instead of `*Response`.

The only things I like about this approach is that it avoids the repeated type assertions of `httpext.Response.Body` we had everywhere, and the silly `processResponse` workaround.

But let me know if anything here is an improvement or could be salvaged. I'm still leaning towards the original suggestion by Mihail to drop both `[]byte` and `ArrayBuffer` support from `res.json()` (and `res.html()`).

Reply:
> Object.keys(resp) still returns only ["body"]

This actually breaks a much more important (IMO) functionality of being able to do:
```javascript
console.log(JSON.stringify(resp, null, "  "));
````
 and getting good debugging information of what all the response fields are ... 

I am of the opinion that one day ... we will need to rewrite this to actually use `DynamicObject`or a better (IMO) Object that we add Define*Property on them, depending on how it turns out) so we can stop doing needless transformations (like this one(It will probably need to have some nonenumerable Symbol property that is the original object so we can get it back).  But that will (as you have noticed) require quite a lot of reworking and likely changes throughout the codebase. Which is one of the reasons I haven't even tried to fix the triple copying of the data that is being uploaded (another problem that will be helped by something like that).

So I am also voting for just dropping the again ... almost certainly totally unused functionality of calling `json()`/`html()` on responses that are binary.

Reply:
Resolving this as `httpext` no longer depends on `goja` (see 6a9ea790).